### DISCUSSION TITLE
Can trivy scan the node_modules installed in /usr/lib using "dnf install", in a container for SBOM

### DISCUSSION THREAD

**Discussion by seth-priya** [aquasecurity/trivy#5074] [Q&A]
### Question

Just to simplify the use case, I have a ubi 8.8 container where I just installed nodejs (and npm). I see these (and many more packages) installed in "/usr/lib/node_modules/" as below

<snip>
.....
/usr/lib/node_modules/npm/node_modules/validate-npm-package-name/package.json
/usr/lib/node_modules/npm/node_modules/walk-up-path/package.json
/usr/lib/node_modules/npm/node_modules/wcwidth/package.json
/usr/lib/node_modules/npm/node_modules/which/package.json
/usr/lib/node_modules/npm/node_modules/wide-align/package.json
/usr/lib/node_modules/npm/node_modules/wrappy/package.json
/usr/lib/node_modules/npm/node_modules/write-file-atomic/package.json
/usr/lib/node_modules/npm/node_modules/yallist/package.json
/usr/lib/node_modules/npm/package.json
</snip>

But now I save this container into an image and run trivy using
trivy image npm-sbom-analysis:latest -f cyclonedx > npm-analysis-test.out

I see no node-pkg in the trivy report, none of the above node_modules showed up.

However, if I install a package using "npm install -g" , trivy is able to scan everything that's in "/usr/local/node-modules", is this expected behaviour?

Is there a way I can list the system / distro installed nodejs packages?

### Target

Container Image

### Scanner

License

### Output Format

CycloneDX

### Mode

Standalone

### Operating System

Linux x86_64

### Version

```bash
[user3@p006vm77 ~]$ trivy --version
Version: 0.36.1
Vulnerability DB:
  Version: 2
  UpdatedAt: 2023-07-18 00:15:41.018069915 +0000 UTC
  NextUpdate: 2023-07-18 06:15:41.018069615 +0000 UTC
  DownloadedAt: 2023-07-18 04:30:48.687375162 +0000 UTC
```

**Comment by seth-priya** [comment]
Team, any inputs on how I can include the nodejs packages installed using yum in the SBOM?

**Reply by knqyf263** [reply]
If they are installed via the OS package manager, they are reported as the `node` package in rpm.

**Reply by seth-priya** [reply]
@knqyf263 thanks for the response, yes, that is true,  but I don't get the dependencies that way, for example, lets say I do a 
"dnf install nodejs", I see all of the below  

[root@9d1998132f89 node_modules]# ls -lrt
total 12
drwxr-xr-x. 3 root root   17 Sep  1 08:40 @iarna
drwxr-xr-x. 2 root root  135 Sep  1 08:40 agent-base
drwxr-xr-x. 2 root root   75 Sep  1 08:40 abbrev
drwxr-xr-x. 3 root root  156 Sep  1 08:40 JSONStream
drwxr-xr-x. 2 root root   79 Sep  1 08:40 ansistyles
drwxr-xr-x. 2 root root   79 Sep  1 08:40 ansicolors
drwxr-xr-x. 2 root root   74 Sep  1 08:40 ansi-styles
drwxr-xr-x. 2 root root   74 Sep  1 08:40 ansi-regex
drwxr-xr-x. 2 root root   94 Sep  1 08:40 ansi-align
drwxr-xr-x. 3 root root  124 Sep  1 08:40 agentkeepalive
drwxr-xr-x. 3 root root  115 Sep  1 08:40 archy
drwxr-xr-x. 2 root root   94 Sep  1 08:40 aproba
drwxr-xr-x. 2 root root  153 Sep  1 08:40 asap
drwxr-xr-x. 3 root root  202 Sep  1 08:40 are-we-there-yet
drwxr-xr-x. 2 root root   93 Sep  1 08:40 assert-plus

and several more of these in /usr/lib/node_modules/npm/node_modules but the only thing the OS package manager would show is npm itself in this context (and of course nodejs and any other dependency installed by dnf itself) but none of the other node packages, is there any config parameter to get these?

**Comment by seth-priya** [comment]
any comments on this? I am trying to use trivy for my use case but this difference (as compared to what's there in the report generated by another team using another tool) is causing a bit of a problem ... 
I understand trivy scans all the stuff under the node_modules for package.json to be able to list them in the SBOM, is there something in the code that explicitly disables checks in the system install path, "/usr/lib/node_modules" or equivalent,  wondering if I can update and rebuild a version of trivy locally to attempt to generate this info, any pointers?

**Comment by seth-priya** [comment]
team, any more inputs on this?

**Comment by iarna** [comment]
Hi, I have nothing to do with this project, but you accidentally tagged me. Anywa, as a former maintainer of npm I maybe might be able to clarify one thing. 

npm (which is the component of nodejs that these modules come from) bundles its dependencies. What that means, is that while npm has a node_modules folder it is better thought of as a statically linked binary than something with runtime dependencies. That is, for the purposes of scanning an install on a system, it and its libraries are a single unit.

As such, it is appropriate that rpm reports it as a single thing. Node projects on the system cannot use those libraries, they can only be used by npm itself,. Flaws found in one of the version of the modules bundled with npm are reported to the npm project itself in addition to the library. If you need to update to avoid an issue that means updating npm itself, not updating any of its dependencies.

(As a side note: Not all vulnerabilities in a dependency of npm will result in a vulnerability _in_ npm: npm may not be using the vulnerable API surface, or the vulnerability may not be relevant to a CLI tool. By treating npm and its dependencies as an atomic unit, security decisions can be made around the actual impact on the project and not merely the impact on a hypothetical project.)

**Comment by seth-priya** [comment]
Thanks @iarna . Firstly apologies, seems like by referencing the "@iarna" package in the above "ls" output snippet , I happended to tag you unknowingly. Secondly, thanks for your detailed and insightful response, it does make a lot of sense to me. The only thing that I was wondering is that since each of these "sub"-packages, is a community of its own, with varied licenses, listing them out would make the SBOM more complete? For the typical C/C++ projects, we usually remove the build/dev dependencies from the container altogether, to, in part, address this issues, but since that's not a possibility here ...

**Reply by AnaisUrlichs** [reply]
Hi @seth-priya just catching up on this conversation, I am not sure I follow the original question. To summarise, if you scan node applications,  you first have to install the packages. Also, when you create an SBOM of your app, the SBOM will have information on the licenses included. Please let me know if there is something that needs additional clarification / that we can help with?

Here is more information https://aquasecurity.github.io/trivy/v0.45/docs/coverage/language/nodejs/

**Comment by seth-priya** [comment]
thanks @AnaisUrlichs I have reviewed the documentation and do have a node application (the original comment in this issue is just a sample / simplified scenario) and the packages that are being installed using the npm package manager in the container and their dependencies are being currently detected for the most part. However, if I simply, lets say, use something like the "find" command in my container to search for all node_modules, I stumble across a whole bunch of these under "/usr/lib/node_modules"  that aren't being captured in the SBOM as they are being installed by the os package manager, in this case, yum/dnf as opposed to one of the nodejs package managers, so was just wondering if there is some mechanism to capture these in the SBOM (as they are actually present in the container) either as dependencies of os packages or npm packages or something else ... or can/should they be completely ignored? would ignoring them be justifiable to the legal team that reviews the license of everything that gets shipped in a container, as an example? Just thinking aloud and appreciate your thoughts, thanks!
